README
ZEsarUX - ZX Second-Emulator And Released for UniX 


It's a ZX Machines Emulator for Unix, including all the Sinclar computers:

-MK14
-ZX80
-ZX81
-ZX Spectrum
-QL
-Z88

And also:

-Timex TS 2068
-Sam Coupe
-Pentagon 
-Chloe 140 SE, 280 SE
-Chrome
-Prism
-ZX-Uno
-ZX-Evolution TS-Conf
-TBBlue/ZX Spectrum Next
-Jupiter Ace
-Amstrad CPC 464

ZEsarUX source code and binaries are distributed under GNU GPL license. 
ZEsarUX also includes third-party roms, media, programs and games NOT covered by this license.

This is my second ZX Spectrum emulator after ZXSpectr
http://sourceforge.net/projects/zxspectr/

I recommend you to read FEATURES, INSTALL and HISTORY files, as well as other documents in this emulator.
You can open them from the help menu or from an external viewer.


ZEsarUX distributed under GNU GENERAL PUBLIC LICENSE v3. You may read it on the LICENSE file.

Includes Musashi 3.4 - A portable Motorola M680x0 processor emulation engine.
Copyright 1998-2002 Karl Stenerud. All rights reserved. Distributed under MIT License, you may read it on the LICENSE_MOTOROLA_CORE file

Includes National Semiconductor SC/MP CPU Emulator
Copyright 2017 Miodrag Milanovic. Distributed under BSD-3 License, you may read it on the LICENSE_SCMP_CORE




Changelog
Version 6.1. 26 January 2018 - Xeno edition

Added BetaDisk and TR-DOS emulation
Added Spectrum +3 emulation
Added Spectrum keyboard ghosting error feature
Added ZX Spectrum Recreated keyboard support
Added setting to exit emulator after a number of seconds (--exit-after)
Added setting to show changelog when updating emulator version
Added actions to file utilities: Copy, Move, Rename
Added setting to write protect IDE, MMC, ZX-Uno Flash
Added setting to disable persistent writes on IDE and MMC
Added two more GUI styles: Clean and CleanInverse
Added second Easter Egg ;)
Added improved Spectrum 16/48/+ colour palette
Added setting to add a watermark on the display
Added setting to reduce display to a factor of 0.75. Used on RunZX event
Added TSConf DMA emulation, line interrupt
Improved joystick support: 
* added joystick test menu
* buttons to events and buttons to joystick configuration are saved on configuration file
Improved OSD keyboard:
* now you can press up to all 40 keys (on Spectrum/ZX80/ZX81) at the same time
* now it can be opened from any place (needs multitask on)
Improved QL emulation: 
* experimental support of LOAD, LRUN command
* can now open qlay mdv microdrive image files
Improved file viewer from file utilities: 
*specific file viewers for these kind of files: .tap, .zx, .zsf, .sp, .z80, .sna, .spg, .p, .o, .mmc/.ide (fat browser, plus3dos browser), .dsk, .trd, .tzx, .cdt, .flash (zxuno flash, z88 flash, superupgrade flash), .epr, .eprom
*default hexadecimal file viewer for unknown files
Improved TSConf rendering: now tiles and sprites are drawn per scanline. Border drawing. Handling of scroll registers
Improved text mode drivers (aalib, stdout, simpletext, curses): text settings are in different submenu, setting to set stdout/simpletext fps
Improved messages windows. You can now have an internal clipboard, press key c to copy to it. Press P on file utils to paste contents to a file
Improved menus, a new "tabbed" menus to improve usability and interaction with mouse, joystick and stdout driver: waveform, visualmem, ay player
Improved menu: now character width can be reduced to 7,6 or 5 pixels width
Improved smartload: now the cursor remembers last file loaded
Improved sprite viewer: 
*now can see sprites with speccy screen format
*you can view sprites up to 256x192 
Improved sound: Silence detector does not activate on pthreads versions
Improved default configuration: when the configuration file doesn't exist, the new one has the autosave configuration setting enabled
Improved configuration parsing: if unknown setting found on configuration file, ZEsarUX will open but warns about the error, and the autosave feature is not enabled
Improved TBBlue emulation: Added clip window registers
Improved rendering on text mode drivers: now realvideo mode can render directly to black&white text. Try TSConf with --vo stdout --realvideo --autoredrawstdout
Fixed TBBlue MMU when using segment 0000-3FFFH
Fixed ZX-Uno emulation: 128kb timings are the same as a Spectrum 128k. When port 1ffd enabled, port FF return FF
Fixed window footer when running Pentagon
Fixed some crashes with speech enabled and opening some menus: AY Registers, AY Player, etc...


Version 6.0. 27 November 2017 - Gunfright edition

Added machine ZX-Evolution TS-Conf
Added machine Chrome (Italian ZX Spectrum Clone)
Added machine MK14
Added lots of new ZEsarUX remote commands (ZRCP). See ZRCP help to know them
Added Visualmem, Poke, Find, Save binary menus also for Sinclair QL
Added QL GUI Style
Added setting to test configuration
Added menu to find and set lives in games
Added menu to view color palettes
Added actions commands to breakpoints, so you can run some actions when a breakpoint is fired
Added up to three AY Chips
Added emulation of other DAC chips appart from Specdrum: Covox, etc
Added Antonio Villena's "Kartusho" cartridge interface
Added quicksave function
Added setting to set 256kb or 512kb on Spectrum 128k machines
Added setting to set F keys (F1, F2, etc) to run different actions: reset, nmi, exit emulator, etc
Added TBBlue features: Sprite chip, Layer2, RTC clock for esxdos, 2 MB RAM, new 8kb block MMU, 9-bit palette, lores mode, layer priorities, partial Next Z80 opcodes (see extras/docs/tbblue/new_opcodes.txt to know which are emulated)
Added ZX-Uno features: Radasoffset and Radaspadding registers, to do hardware scrolling
Added ESXDOS handler so file operations can be done with files stored on your computer hard disk, and not only on virtual mmc/ide files. Examples can be found on extras/media/spectrum/esxdos_handler/. "Classic" mmc/ide emulation by using virtual mmc/ide files can also be used
Added feature "memory zone" to ZRCP and all menus that inspect memory (debug, disassemble, sprites etc): now you can look at all memory zones, for example: full 128kb ram, 64kb rom of +2A, divmmc ram, etc
Added file utils menu, to view files, delete, truncate, etc
Added boot logo
Added setting to allow to load a .sna snapshot file without changing the current machine type
Added setting to allow to truncate a tape file when selecting it as tape out file

Moved folders: docs, media, snap_tests to a "ZEsarUX-extras" package. These folders have been separated from the emulator main distributed file, and the main download have been reduced to 5 MB aprox. The git repository still holds all the files (extras have been moved to folder "extras"), but maybe in the future I will remove them also from the repo.
Moved mmc/ide files location to extras/media/disk_images folder

Improved menu interface:
*menu gui is zoomed when running following machines: QL, TSConf, CPC, Prism, Sam
*menu movement with mouse
*colours are greyscaled when multitask disabled
*allow blinking text, like the input cursor (needs multitask on)
*new splash window type which disappears automatically (needs multitask on). Used only on change memory zones from debug cpu menu, but it will be used on more places probably

Improved debugger:

cpu.h
/*
    ZEsarUX  ZX Second-Emulator And Released for UniX
    Copyright (C) 2013 Cesar Hernandez Bano

    This file is part of ZEsarUX.

    ZEsarUX is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef CPU_H
#define CPU_H

#include "compileoptions.h"

/*
Emulator version format:
XX.YY-S

XX: Major version number. 1 or 2 digits number. Currently using only one digit but could be greater (on ZEsarUX version 10 or higher)
YY: Minor version number. 1 or 2 digits number
-S: Optional string, only present on non-stable versions. S it's normally "SN" or "RC", where RC can also be RCX, being X a digit: RC1, RC2, etc

Examples

6.0: Means major version 6, minor version 0. It's a stable version (no suffix string -....)

6.1-RC: Means major version 6, minor version 1, beta version: RC
6.2-RC3: Means major version 6, minor version 2, beta version: RC3

*/

//#define EMULATOR_VERSION "6.1"

//#define EMULATOR_VERSION "6.1-RC"
#define EMULATOR_VERSION "6.1-SN"
#define SNAPSHOT_VERSION

#define EMULATOR_DATE "26 January 2018"
#define EMULATOR_SHORT_DATE "26/01/2018"
#define EMULATOR_GAME_EDITION "Xeno"
#define EMULATOR_EDITION_NAME EMULATOR_GAME_EDITION " edition"

//8 bits
typedef unsigned char z80_byte;
//16 bits
typedef unsigned short z80_int;
//32 bits
typedef unsigned int z80_long_int;

//Usado en menu, para registros de 16 bits de z80 y de 32 de motorola
typedef unsigned int menu_z80_moto_int;

struct s_z80_bit {
	unsigned int v:1;
};


typedef union {
#ifdef WORDS_BIGENDIAN
  struct { z80_byte h,l; } b;
#else
  struct { z80_byte l,h; } b;
#endif

  z80_int w;
} z80_registro;


typedef struct s_z80_bit z80_bit;


extern int zoom_x,zoom_y;
extern int zoom_x_original,zoom_y_original;


extern z80_registro registro_hl;
#define reg_h registro_hl.b.h
#define reg_l registro_hl.b.l
#define HL registro_hl.w
#define reg_hl registro_hl.w


extern z80_registro registro_de;
#define reg_d registro_de.b.h
#define reg_e registro_de.b.l
#define DE registro_de.w
#define reg_de registro_de.w


extern z80_registro registro_bc;
#define reg_b registro_bc.b.h
#define reg_c registro_bc.b.l
#define BC registro_bc.w
#define reg_bc registro_bc.w


extern z80_byte reg_i;
extern z80_byte reg_r,reg_r_bit7;

#define IR ( ( reg_i ) << 8 | ( reg_r_bit7 & 0x80 ) | ( reg_r & 0x7f ) )

extern z80_byte reg_h_shadow,reg_l_shadow;
extern z80_byte reg_b_shadow,reg_c_shadow;
extern z80_byte reg_d_shadow,reg_e_shadow;
extern z80_byte reg_a_shadow;

#define REG_AF (value_8_to_16(reg_a,Z80_FLAGS))

#define REG_AF_SHADOW (value_8_to_16(reg_a_shadow,Z80_FLAGS_SHADOW))
#define REG_HL_SHADOW (value_8_to_16(reg_h_shadow,reg_l_shadow))
#define REG_BC_SHADOW (value_8_to_16(reg_b_shadow,reg_c_shadow))
#define REG_DE_SHADOW (value_8_to_16(reg_d_shadow,reg_e_shadow))

extern void set_machine(char *romfile);
extern void set_machine_params(void);
extern void post_set_machine(char *romfile);
extern void post_set_machine_no_rom_load(void);

extern z80_int reg_pc;
extern z80_byte reg_a;
extern z80_int reg_sp;
extern z80_int reg_ix;
extern z80_int reg_iy;

extern z80_byte Z80_FLAGS;
extern z80_byte Z80_FLAGS_SHADOW;

#define FLAG_C  0x01
#define FLAG_N  0x02
#define FLAG_PV 0x04
#define FLAG_3  0x08
#define FLAG_H  0x10
#define FLAG_5  0x20
#define FLAG_Z  0x40
#define FLAG_S  0x80

extern z80_int memptr;

extern z80_bit iff1,iff2;

extern z80_bit interrupcion_pendiente;
extern z80_bit z80_ejecutando_halt;
extern z80_byte im_mode;
extern z80_bit cpu_step_mode;

#ifndef GCC_UNUSED

#ifdef __GNUC__
#  define GCC_UNUSED __attribute__((unused))
#else
#  define GCC_UNUSED
#endif

#endif


extern z80_byte puerto_65278; //    db    255  ; V    C    X    Z    Sh    ;0
extern z80_byte puerto_65022; //    db    255  ; G    F    D    S    A     ;1
extern z80_byte puerto_64510; //    db              255  ; T    R    E    W    Q     ;2
extern z80_byte puerto_63486; //    db              255  ; 5    4    3    2    1     ;3
extern z80_byte puerto_61438; //    db              255  ; 6    7    8    9    0     ;4
extern z80_byte puerto_57342; //    db              255  ; Y    U    I    O    P     ;5
extern z80_byte puerto_49150; //    db              255  ; H                J         K      L    Enter ;6
extern z80_byte puerto_32766; //    db              255  ; B    N    M    Simb Space ;7
extern z80_byte puerto_especial1;
extern z80_byte puerto_especial2;
extern z80_byte puerto_especial3;
extern z80_byte puerto_especial4;


extern z80_int *registro_ixiy;

extern z80_bit border_enabled;


extern void (*codsinpr[]) ();
extern void (*codprcb[]) ();
extern void (*codpred[]) ();

extern void (*codprddfd[]) ();

extern z80_byte *memoria_spectrum;

//extern z80_byte inves_ula_delay_factor;
extern z80_byte current_machine_type;
extern z80_bit modificado_border;

extern z80_bit inves_ula_bright_error;




extern void set_undocumented_flags_bits(z80_byte value);
extern void set_flags_zero_sign(z80_byte value);
extern void set_flags_zero_sign_16(z80_int value);
extern void set_flags_carry_suma(z80_byte antes,z80_byte result);
extern void set_flags_carry_resta(z80_byte antes,z80_byte result);
extern void set_flags_carry_16_suma(z80_int antes,z80_int result);
extern void set_flags_carry_16_resta(z80_int antes,z80_int result);

extern void reset_cpu(void);
extern void cold_start_cpu_registers(void);

#define value_8_to_16(h,l) (((h)<<8)|l)

#define value_16_to_8l(hl) ((hl) & 0xFF)
#define value_16_to_8h(hl) (((hl)>>8) & 0xFF)

extern z80_byte *devuelve_reg_offset(z80_byte valor);

extern int video_zx8081_linecntr;
extern z80_bit video_zx8081_linecntr_enabled;

extern z80_byte video_zx8081_ula_video_output;


//#define TEMP_LINEAS_ZX8081 192
//#define TEMP_BORDE_SUP_ZX8081 56


//#define macro_invalid_opcode(MESSAGE) debug_printf(VERBOSE_ERR,"Invalid opcode " MESSAGE ". Final PC: %X",reg_pc)

//extern z80_bit debug_cpu_core_loop;
//extern int cpu_core_loop_active;
//extern void (*cpu_core_loop) (void);

#define CPU_CORE_SPECTRUM 1
#define CPU_CORE_ZX8081 2
#define CPU_CORE_Z88 3
#define CPU_CORE_ACE 4
#define CPU_CORE_CPC 5
#define CPU_CORE_SAM 6
#define CPU_CORE_QL 7
#define CPU_CORE_MK14 8

extern struct timeval z80_interrupts_timer_antes, z80_interrupts_timer_ahora;
extern long z80_timer_difftime, z80_timer_seconds, z80_timer_useconds;
extern struct timeval zesarux_start_time;
extern z80_bit interrupcion_timer_generada;
extern int contador_frames_veces_buffer_audio;
extern z80_bit esperando_tiempo_final_t_estados;
extern int debug_registers;
extern z80_bit interrupcion_maskable_generada;
extern z80_bit interrupcion_non_maskable_generada;
extern z80_bit interrupcion_timer_generada;
extern z80_byte reg_r_antes_zx8081;
extern z80_bit temp_zx8081_lineasparimpar;
extern char *get_machine_name(z80_byte m);

extern int porcentaje_velocidad_emulador;
extern void set_emulator_speed(void);

//T-estados totales del frame
extern int t_estados;

extern int t_scanline;

//Scan line actual
extern int t_scanline_draw;
extern int t_scanline_draw_timeout;

//Autoseleccionar opciones de emulacion (audio, realvideo, etc) segun snap o cinta cargada
extern z80_bit autoselect_snaptape_options;

extern z80_bit tape_loading_simulate;
extern z80_bit tape_loading_simulate_fast;
extern void end_emulator(void);

extern z80_bit snow_effect_enabled;

struct s_driver_struct
{
        char driver_name[30];
        int (*funcion_init) () ;
        int (*funcion_set) () ;
};

typedef struct s_driver_struct driver_struct;

extern driver_struct scr_driver_array[];
extern int num_scr_driver_array;

extern driver_struct audio_driver_array[];
extern int num_audio_driver_array;

#define MAX_SCR_INIT 10
#define MAX_AUDIO_INIT 10

extern z80_bit stdout_simpletext_automatic_redraw;
//Valores para stdout. Estan aqui porque se graban en archivo .zx, aunque no este el driver stdout compilado
//fin valores para stdout


#define MACHINE_ID_SPECTRUM_16			0
#define MACHINE_ID_SPECTRUM_48			1
#define MACHINE_ID_INVES			2
#define MACHINE_ID_MICRODIGITAL_TK90X		3
#define MACHINE_ID_MICRODIGITAL_TK90X_SPA	4
#define MACHINE_ID_MICRODIGITAL_TK95		5
#define MACHINE_ID_SPECTRUM_128			6
#define MACHINE_ID_SPECTRUM_128_SPA		7
#define MACHINE_ID_SPECTRUM_P2			8
#define MACHINE_ID_SPECTRUM_P2_FRE		9
#define MACHINE_ID_SPECTRUM_P2_SPA		10
#define MACHINE_ID_SPECTRUM_P2A_40		11
#define MACHINE_ID_SPECTRUM_P2A_41		12
#define MACHINE_ID_SPECTRUM_P2A_SPA		13
#define MACHINE_ID_ZXUNO			14
#define MACHINE_ID_CHLOE_140SE			15
#define MACHINE_ID_CHLOE_280SE			16
#define MACHINE_ID_TIMEX_TS2068			17
#define MACHINE_ID_PRISM			18
#define MACHINE_ID_TBBLUE			19
#define MACHINE_ID_SPECTRUM_48_SPA		20

#define MACHINE_ID_PENTAGON			21
#define MACHINE_ID_CHROME			22

#define MACHINE_ID_TSCONF			23
#define MACHINE_ID_BASECONF			24


#define MACHINE_ID_SPECTRUM_P3_40		25
#define MACHINE_ID_SPECTRUM_P3_41		26
#define MACHINE_ID_SPECTRUM_P3_SPA		27


#define MACHINE_ID_ZX80				120
#define MACHINE_ID_ZX81				121
#define MACHINE_ID_ACE				122
#define MACHINE_ID_Z88				130

#define MACHINE_ID_CPC_464			140
#define MACHINE_ID_SAM				150
#define MACHINE_ID_QL_STANDARD			160
#define MACHINE_ID_MK14_STANDARD		180



//Condiciones de maquinas activas
#define MACHINE_IS_SPECTRUM (current_machine_type<30)
#define MACHINE_IS_SPECTRUM_16 (current_machine_type==MACHINE_ID_SPECTRUM_16)
#define MACHINE_IS_SPECTRUM_48 (current_machine_type==MACHINE_ID_SPECTRUM_48)
#define MACHINE_IS_INVES (current_machine_type==MACHINE_ID_INVES)
#define MACHINE_IS_SPECTRUM_48_SPA (current_machine_type==MACHINE_ID_SPECTRUM_48_SPA)
#define MACHINE_IS_SPECTRUM_16_48 ( (current_machine_type<=MACHINE_ID_MICRODIGITAL_TK95) || MACHINE_IS_SPECTRUM_48_SPA )

#define MACHINE_IS_SPECTRUM_128_P2 ( (current_machine_type>=MACHINE_ID_SPECTRUM_128 && current_machine_type<=MACHINE_ID_SPECTRUM_P2_SPA) || MACHINE_IS_PENTAGON)

#define MACHINE_IS_SPECTRUM_P3 (current_machine_type==MACHINE_ID_SPECTRUM_P3_40 || current_machine_type==MACHINE_ID_SPECTRUM_P3_41 || current_machine_type==MACHINE_ID_SPECTRUM_P3_SPA)

//Incluir en MACHINE_IS_SPECTRUM_P2A los P3, pues son lo mismo pero con disquetera
#define MACHINE_IS_SPECTRUM_P2A ( (current_machine_type>=MACHINE_ID_SPECTRUM_P2A_40 && current_machine_type<=MACHINE_ID_SPECTRUM_P2A_SPA) || MACHINE_IS_SPECTRUM_P3)
#define MACHINE_IS_SPECTRUM_128_P2_P2A ( MACHINE_IS_SPECTRUM_128_P2 || MACHINE_IS_SPECTRUM_P2A)

#define MACHINE_IS_ZXUNO (current_machine_type==MACHINE_ID_ZXUNO)

#define MACHINE_IS_CHLOE_140SE (current_machine_type==MACHINE_ID_CHLOE_140SE)
#define MACHINE_IS_CHLOE_280SE (current_machine_type==MACHINE_ID_CHLOE_280SE)
#define MACHINE_IS_CHLOE (MACHINE_IS_CHLOE_140SE || MACHINE_IS_CHLOE_280SE)
#define MACHINE_IS_TIMEX_TS2068 (current_machine_type==MACHINE_ID_TIMEX_TS2068)
#define MACHINE_IS_PRISM (current_machine_type==MACHINE_ID_PRISM)
#define MACHINE_IS_TBBLUE (current_machine_type==MACHINE_ID_TBBLUE)
#define MACHINE_IS_PENTAGON (current_machine_type==MACHINE_ID_PENTAGON)
#define MACHINE_IS_CHROME (current_machine_type==MACHINE_ID_CHROME)


#define MACHINE_IS_TSCONF (current_machine_type==MACHINE_ID_TSCONF)
#define MACHINE_IS_BASECONF (current_machine_type==MACHINE_ID_BASECONF)

#define MACHINE_IS_EVO (MACHINE_IS_TSCONF || MACHINE_IS_BASECONF)

#define MACHINE_IS_ZX80 (current_machine_type==MACHINE_ID_ZX80)
#define MACHINE_IS_ZX81 (current_machine_type==MACHINE_ID_ZX81)
#define MACHINE_IS_ZX8081 (MACHINE_IS_ZX80 || MACHINE_IS_ZX81)
#define MACHINE_IS_ACE (current_machine_type==MACHINE_ID_ACE)
#define MACHINE_IS_ZX8081ACE (MACHINE_IS_ZX8081 || MACHINE_IS_ACE)

#define MACHINE_IS_Z88 (current_machine_type==MACHINE_ID_Z88)

#define MACHINE_IS_CPC_464 (current_machine_type==MACHINE_ID_CPC_464)
#define MACHINE_IS_CPC (current_machine_type>=MACHINE_ID_CPC_464 && current_machine_type<=149)
#define MACHINE_IS_SAM (current_machine_type==MACHINE_ID_SAM)

/*
160=QL
161-179 reservado para otros QL
*/

#define MACHINE_IS_QL_STANDARD (current_machine_type==MACHINE_ID_QL_STANDARD)
#define MACHINE_IS_QL (current_machine_type>=MACHINE_ID_QL_STANDARD && current_machine_type<=179)

/*
180=MK14 Standard
181-189 reservado para otros MK14
*/

#define MACHINE_IS_MK14_STANDARD (current_machine_type==MACHINE_ID_MK14_STANDARD)
#define MACHINE_IS_MK14 (current_machine_type>=MACHINE_ID_MK14_STANDARD && current_machine_type<=189)


#define CPU_IS_MOTOROLA (MACHINE_IS_QL)
#define CPU_IS_SCMP (MACHINE_IS_MK14)
#define CPU_IS_Z80 (!CPU_IS_MOTOROLA && !CPU_IS_SCMP)


extern int machine_emulate_memory_refresh;
extern int machine_emulate_memory_refresh_counter;

extern z80_byte last_inves_low_ram_poke_menu;

extern void random_ram_inves(z80_byte *puntero,int longitud);



//valor obtenido probando
#define MAX_EMULATE_MEMORY_REFRESH_COUNTER 1500000
#define MAX_EMULATE_MEMORY_REFRESH_LIMIT (MAX_EMULATE_MEMORY_REFRESH_COUNTER/2)


extern int zesarux_main (int main_argc,char *main_argv[]);
extern z80_bit no_cambio_parametros_maquinas_lentas;
extern z80_bit opcion_no_splash;

extern int argc;
extern char **argv;
extern int puntero_parametro;
extern int siguiente_parametro(void);

extern void siguiente_parametro_argumento(void);

extern int joystickkey_definidas;
extern void rom_load(char *romfilename);
extern void hard_reset_cpu(void);

extern z80_bit zxmmc_emulation;

extern z80_bit quickexit;

extern z80_bit azerty_keyboard;

extern z80_int ramtop_ace;

extern z80_bit allow_write_rom;

extern int z88_cpc_keymap_type;

extern z80_bit chloe_keyboard;

extern unsigned int debug_t_estados_parcial;

#define MAX_CPU_TURBO_SPEED 16

extern int cpu_turbo_speed;

extern void cpu_set_turbo_speed(void);

extern z80_bit windows_no_disable_console;

extern char *string_machines_list_description;

extern void set_menu_gui_zoom(void);


extern int exit_emulator_after_seconds;

extern int exit_emulator_after_seconds_counter;

extern z80_bit do_no_show_changelog_when_update;


#endif
